File: app/src/test/java/com/chefvoice/app/util/ShoppingListTest.kt
package com.chefvoice.app.util

import com.chefvoice.app.model.Ingredient
import com.chefvoice.app.model.ShoppingItem
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * A shopping list with a duplicate on it is a small annoyance. One with a silently
 * wrong total sends the chef to the shop to buy the wrong amount, and they will not
 * find out until they are cooking. So these tests pin the merge as deliberately
 * narrow: combine only when it is certainly safe, and split into two lines otherwise.
 */
class ShoppingListTest {
    private fun item(quantity: String, unit: String, name: String, from: String = "Chili") =
        ShoppingItem(quantity = quantity, unit = unit, name = name, recipeTitle = from)

    @Test
    fun addsQuantitiesForTheSameThingInTheSameUnit() {
        val merged = ShoppingList.merge(
            listOf(item("2", "cups", "flour")),
            listOf(item("1", "cup", "flour"))
        )
        assertEquals(1, merged.size)
        assertEquals("3", merged.single().quantity)
    }

    @Test
    fun foldsUnitSpellingsOntoOneLine() {
        val merged = ShoppingList.merge(
            listOf(item("2", "tablespoons", "olive oil")),
            listOf(item("1", "tbsp", "olive oil"))
        )
        assertEquals(1, merged.size)
        assertEquals("3", merged.single().quantity)
    }

    @Test
    fun matchesSingularAndPluralIngredientNames() {
        val merged = ShoppingList.merge(
            listOf(item("2", "", "onions")),
            listOf(item("1", "", "Onion"))
        )
        assertEquals(1, merged.size)
        assertEquals("3", merged.single().quantity)
    }

    @Test
    fun neverAddsAcrossIncomparableUnits() {
        // Turning 200 g of flour into cups needs to know how flour packs. Two lines
        // is the honest answer.
        val merged = ShoppingList.merge(
            listOf(item("2", "cups", "flour")),
            listOf(item("200", "g", "flour"))
        )
        assertEquals(2, merged.size)
    }

    @Test
    fun keepsDifferentIngredientsApart() {
        val merged = ShoppingList.merge(
            listOf(item("2", "cups", "flour")),
            listOf(item("2", "cups", "sugar"))
        )
        assertEquals(2, merged.size)
    }

    @Test
    fun doesNotMergeWhenEitherQuantityCannotBeRead() {
        // Adding "a pinch" to "1 tsp" would mean inventing a number.
        val merged = ShoppingList.merge(
            listOf(item("a pinch", "", "salt")),
            listOf(item("1", "tsp", "salt"))
        )
        assertEquals(2, merged.size)
    }

    @Test
    fun aMergedLineRemembersEveryRecipeItCameFrom() {
        val merged = ShoppingList.merge(
            listOf(item("2", "cups", "flour", from = "Chili")),
            listOf(item("1", "cup", "flour", from = "Cornbread"))
        )
        assertEquals("Chili, Cornbread", merged.single().recipeTitle)
    }

    @Test
    fun mergingIntoACheckedLineBringsItBackOntoTheList() {
        // The chef already bought 2 cups and ticked it off. Adding a recipe that needs
        // another cup means there is more to buy, so the line must not stay satisfied.
        val bought = item("2", "cups", "flour").copy(checked = true)
        val merged = ShoppingList.merge(listOf(bought), listOf(item("1", "cup", "flour")))
        assertEquals(1, merged.size)
        assertFalse(merged.single().checked)
    }

    @Test
    fun buildingFromARecipeCarriesTheScaling() {
        // A recipe doubled on screen must buy twice as much.
        val items = ShoppingList.itemsFor(
            recipeId = "r1",
            recipeTitle = "Chili",
            ingredients = listOf(Ingredient(quantity = "2", unit = "cups", name = "beans")),
            servingFactor = 2.0
        )
        assertEquals("4", items.single().quantity)
        assertEquals("r1", items.single().recipeId)
        assertEquals("Chili", items.single().recipeTitle)
    }

    @Test
    fun buildingFromARecipeDropsOnlyNamelessLines() {
        val items = ShoppingList.itemsFor(
            recipeId = "r1",
            recipeTitle = "Chili",
            ingredients = listOf(
                Ingredient(quantity = "2", unit = "cups", name = "beans"),
                Ingredient(quantity = "1", unit = "tsp", name = "  "),
                Ingredient(quantity = "", unit = "", name = "black pepper")
            )
        )
        // The blank-name row has nothing to buy; the quantity-less one still does.
        assertEquals(2, items.size)
        assertTrue(items.any { it.name == "black pepper" })
    }

    @Test
    fun shareTextMarksWhatIsAlreadyBought() {
        val text = ShoppingList.asShareText(
            listOf(item("2", "cups", "flour"), item("1", "tsp", "salt").copy(checked = true))
        )
        assertTrue(text.contains("[ ] 2 cups flour"))
        assertTrue(text.contains("[x] 1 tsp salt"))
    }

    @Test
    fun shareTextOfAnEmptyListSaysSo() {
        assertTrue(ShoppingList.asShareText(emptyList()).contains("empty"))
    }
}

File: web/tests/shopping-list.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { merge, itemsFor, asShareText } from '../js/shopping-list.js';

// Port of ShoppingListTest.kt. A duplicate is a small annoyance; a silently wrong total sends
// the chef to buy the wrong amount, so the merge is deliberately narrow.

const item = (quantity, unit, name, from = 'Chili') => ({ id: `${name}${quantity}${unit}`, quantity, unit, name, recipeTitle: from, checked: false });

test('adds quantities for the same thing in the same unit', () => {
  const m = merge([item('2', 'cups', 'flour')], [item('1', 'cup', 'flour')]);
  assert.equal(m.length, 1);
  assert.equal(m[0].quantity, '3');
});
test('folds unit spellings onto one line', () => {
  const m = merge([item('2', 'tablespoons', 'olive oil')], [item('1', 'tbsp', 'olive oil')]);
  assert.equal(m.length, 1);
  assert.equal(m[0].quantity, '3');
});
test('matches singular and plural names', () => {
  const m = merge([item('2', '', 'onions')], [item('1', '', 'Onion')]);
  assert.equal(m.length, 1);
  assert.equal(m[0].quantity, '3');
});
test('never adds across incomparable units', () => {
  assert.equal(merge([item('2', 'cups', 'flour')], [item('200', 'g', 'flour')]).length, 2);
});
test('keeps different ingredients apart', () => {
  assert.equal(merge([item('2', 'cups', 'flour')], [item('2', 'cups', 'sugar')]).length, 2);
});
test('does not merge when either quantity cannot be read', () => {
  assert.equal(merge([item('a pinch', '', 'salt')], [item('1', 'tsp', 'salt')]).length, 2);
});
test('a merged line remembers every recipe it came from', () => {
  const m = merge([item('2', 'cups', 'flour', 'Chili')], [item('1', 'cup', 'flour', 'Cornbread')]);
  assert.equal(m[0].recipeTitle, 'Chili, Cornbread');
});
test('merging into a checked line brings it back onto the list', () => {
  const bought = { ...item('2', 'cups', 'flour'), checked: true };
  const m = merge([bought], [item('1', 'cup', 'flour')]);
  assert.equal(m.length, 1);
  assert.equal(m[0].checked, false);
});
test('building from a recipe carries the scaling', () => {
  const items = itemsFor('r1', 'Chili', [{ quantity: '2', unit: 'cups', name: 'beans' }], 2);
  assert.equal(items[0].quantity, '4');
  assert.equal(items[0].recipeId, 'r1');
  assert.equal(items[0].recipeTitle, 'Chili');
});
test('building from a recipe drops only nameless lines', () => {
  const items = itemsFor('r1', 'Chili', [
    { quantity: '2', unit: 'cups', name: 'beans' },
    { quantity: '1', unit: 'tsp', name: '  ' },
    { quantity: '', unit: '', name: 'black pepper' }
  ]);
  assert.equal(items.length, 2);
  assert.ok(items.some(i => i.name === 'black pepper'));
});
test('share text marks what is already bought', () => {
  const text = asShareText([item('2', 'cups', 'flour'), { ...item('1', 'tsp', 'salt'), checked: true }]);
  assert.ok(text.includes('[ ] 2 cups flour'));
  assert.ok(text.includes('[x] 1 tsp salt'));
});
test('share text of an empty list says so', () => assert.ok(asShareText([]).includes('empty')));
